src/ipc_cache.c

```diff
@@ -0,0 +1,225 @@
+        umf_ba_global_free(cache_global);
+        return UMF_RESULT_ERROR_UNKNOWN;
+    }
+
+    cache_global->cache_allocator =
+        umf_ba_create(sizeof(ipc_handle_cache_entry_t));
+    if (!cache_global->cache_allocator) {
+        LOG_ERR("Failed to create IPC cache allocator");
+        umf_ba_global_free(cache_global);
+        return UMF_RESULT_ERROR_OUT_OF_HOST_MEMORY;
+    }
+
+    cache_global->max_size = 0;
+    cache_global->cur_size = 0;
+    cache_global->lru_list = NULL;
+
+    IPC_MAPPED_CACHE_GLOBAL = cache_global;
+
+    return UMF_RESULT_SUCCESS;
+}
+
+static size_t getGlobalLruListSize(lru_list_t lru_list) {
+    size_t size = 0;
+    ipc_handle_cache_entry_t *tmp;
+    DL_COUNT(lru_list, tmp, size);
+    return size;
+}
+
+void umfIpcCacheGlobalTearDown(void) {
+    ipc_mapped_handle_cache_global_t *cache_global = IPC_MAPPED_CACHE_GLOBAL;
+    IPC_MAPPED_CACHE_GLOBAL = NULL;
+
+    if (!cache_global) {
+        return;
+    }
+
+    assert(cache_global->cur_size == 0);
+    assert(getGlobalLruListSize(cache_global->lru_list) == 0);
+
+    umf_ba_destroy(cache_global->cache_allocator);
+    umf_ba_global_free(cache_global);
+}
+
+ipc_mapped_handle_cache_handle_t umfIpcHandleMappedCacheCreate(
+    ipc_mapped_handle_cache_eviction_cb_t eviction_cb) {
+    if (eviction_cb == NULL) {
+        LOG_ERR("Eviction callback is NULL");
+        return NULL;
+    }
+
+    ipc_mapped_handle_cache_t *cache = umf_ba_global_alloc(sizeof(*cache));
+
+    if (!cache) {
+        LOG_ERR("Failed to allocate memory for the IPC cache");
+        return NULL;
+    }
+
+    assert(IPC_MAPPED_CACHE_GLOBAL != NULL);
+
+    cache->global = IPC_MAPPED_CACHE_GLOBAL;
+    cache->hash_table = NULL;
+    cache->eviction_cb = eviction_cb;
+
+    return cache;
+}
+
+void umfIpcHandleMappedCacheDestroy(ipc_mapped_handle_cache_handle_t cache) {
+    ipc_handle_cache_entry_t *entry, *tmp;
+    HASH_ITER(hh, cache->hash_table, entry, tmp) {
+        DL_DELETE(cache->global->lru_list, entry);
+        HASH_DEL(cache->hash_table, entry);
+        cache->global->cur_size -= 1;
+        cache->eviction_cb(&entry->key, &entry->value);
+        umf_ba_free(cache->global->cache_allocator, entry);
+    }
+    HASH_CLEAR(hh, cache->hash_table);
+
+    umf_ba_global_free(cache);
+}
+
+umf_result_t
+umfIpcHandleMappedCacheGet(ipc_mapped_handle_cache_handle_t cache,
+                           const ipc_mapped_handle_cache_key_t *key,
+                           uint64_t handle_id,
+                           ipc_mapped_handle_cache_value_t **retEntry) {
+    ipc_handle_cache_entry_t *entry = NULL;
+    umf_result_t ret = UMF_RESULT_SUCCESS;
+    bool evicted = false;
+    ipc_mapped_handle_cache_value_t evicted_value;
+
+    if (!cache || !key || !retEntry) {
+        LOG_ERR("Some arguments are NULL, cache=%p, key=%p, retEntry=%p",
+                (void *)cache, (const void *)key, (void *)retEntry);
```

Comment:
do we need those casts?

Reply:
Yeah, otherwise there is a compiler warning, like this
```
format ‘%p’ expects argument of type ‘void *’, but argument 4 has type ‘ipc_mapped_handle_cache_handle_t’ {aka ‘struct ipc_mapped_handle_cache_t *’} [-Wformat=]
```